feat: add hooks detail (#6)
kawamurakazushi/presentation-react-hooks: slides.md

import React, { useState } from "react";

const App = () => {
  const [name, setName] = useState("Kawamura");
  const [favoriteFood, setFavFoodName] = useState("curry");

  const handleNameChange = e => {
    setName(e.target.value);
  };

>   const handleFavFoodNameChange = e => {

Reviewer: ```suggestion
  const handleFavFoodChange = e => {
```

    setFavFoodName(e.target.value);
  };
  return (
    <>
      <input value={name} onChange={handleNameChange} />

>       <input value={favoriteFood} onChange={handleFavFoodChange} />

Reviewer: ```suggestion
      <input value={favoriteFood} onChange={handleFavFoodNameChange} />
```

    </>
  );
};
```

---

### Hooks - State

**useState**

```typescript
import { useState } from "react";

function Example() {
  // Declare a new state variable, which we'll call "count"
  const [count, setCount] = useState(0);

  return (
    <div>
      <p>You clicked {count} times</p>
      <button onClick={() => setCount(count + 1)}>Click me</button>
    </div>
  );
}
```

- Called inside a function component to add _local state_.
- React will preserve this state between re-renders.
- `useState` returns a pair: the current state value and a function that lets you update the state.

> The first argument the _initial state_.  

Reviewer: ```suggestion
The first argument is the _initial state_.  
```

Unlike `this.state`, the state here doesn’t have to be an object.

> <!-- So now we can manupulate the react local state from a functinal component  -->

Reviewer: ```suggestion
<!-- So now we can manipulate the react local state from a functional component  -->
```

```typescript
import React, { useContext } from "react";
import { ThemeContext, LocaleContext } from "./myContexts";

const App = () => {
  const theme = useContext(ThemeContext);
  const locale = useContext(LocaleContext);

  return <div className={theme}>{locale}</div>;
};
```

---

### Example - Without Hook - Effects

<!-- Next, what if we want to use some lifecycle methods., Let's see what it will be like in the class Component. -->
<!-- You’ve likely performed data fetching, subscriptions, or manually changing the DOM from React components before. We call these operations “side effects” (or “effects” for short) because they can affect other components and can’t be done during rendering. -->

What happens if you want to use the lifecyle method from the React.Component class?

- Data fetching, subscriptions, or manually changing the DOM from React components.

```typescript
class App extends React.Component {
  componentDidMount() {
    API.subscribeToMessage(this.props.id);
  }
  componentWillUpdate() {
    API.subscribeToMessage(this.props.id);
  }
  componentWillMount() {
    API.unsubscribeToMessage(this.props.id);
  }
  render() {

>     return <></>;

Reviewer: ```suggestion
    return; // UI
```

  }
}
```

---

### Example - With Hook - Effect

**useEffect**

```typescript
import React, { useEffect } from "react";

const App = props => {
  useEffect(
    () => {
      API.subscribeToMessage(props.id);

      return () => {
        API.unSubscribeToMessage(props.id);
      };
    },
    [props.id] // Only re-subscribe if props.id changes
  );

>   return <UI />;

Reviewer: ```suggestion
  return; // UI
```

};
```

- First argument is a `function` that will be triggered after performing the DOM updates. (componentDidMount, componentWillUpdate)

> - You can tell React to skip applying an effect if certain values haven’t changed between re-renders. To do so, pass an array as an optional second argument to `useEffect`:

Reviewer: Maybe we should say that the second argument is called "inputs"
